share/dkg/dkg.go

```diff
@@ -0,0 +1,611 @@
+		DealerIndex: cc.DealerIndex,
+		Share:       deal.SecShare,
+	}
+
+	msg := rc.Hash(d.suite)
+	var err error
+	rc.Signature, err = sign.Schnorr(d.suite, d.long, msg)
+	if err != nil {
+		return nil, err
+	}
+	d.pendingReconstruct[cc.DealerIndex] = append(d.pendingReconstruct[cc.DealerIndex], rc)
+	return rc, nil
+}
+
+// ProcessReconstructCommits takes a ReconstructCommits message and stores it
+// along any others. If there is enough message to recover the coefficients of
+// the public polynomials of the malicious dealer in question, then the
+// polynomial is recovered.
+func (d *DistKeyGenerator) ProcessReconstructCommits(rs *ReconstructCommits) error {
+	if _, ok := d.reconstructed[rs.DealerIndex]; ok {
+		// commitments already reconstructed, no need for other shares
+		return nil
+	}
+	_, ok := d.commitments[rs.DealerIndex]
+	if ok {
+		return errors.New("dkg: commitments not invalidated by any complaints")
+	}
+
+	pub, ok := findPub(d.participants, rs.Index)
+	if !ok {
+		return errors.New("dkg: reconstruct commits with invalid verifier index")
+	}
+
+	msg := rs.Hash(d.suite)
+	if err := sign.VerifySchnorr(d.suite, pub, msg, rs.Signature); err != nil {
+		return err
+	}
+
+	var arr = d.pendingReconstruct[rs.DealerIndex]
+	// check if packet is already received or not
+	for _, r := range arr {
+		if r.Index == rs.Index {
+			// no-op ? error ?
+			return nil
+		}
+	}
+	// add it to list of pending
+	arr = append(arr, rs)
+	d.pendingReconstruct[rs.DealerIndex] = arr
+	// check if we can reconstruct commitments
+	if len(arr) >= d.t {
+		var shares = make([]*share.PriShare, 0, len(arr))
+		for _, r := range arr {
+			shares = append(shares, r.Share)
+		}
+		// error only happens when you have less than t shares, but we ensure
+		// there are more just before
+		pri, _ := share.RecoverPriPoly(d.suite, shares, d.t, len(d.participants))
+		d.commitments[rs.DealerIndex] = pri.Commit(d.suite.Point().Base())
+		// note it has been reconstructed.
+		d.reconstructed[rs.DealerIndex] = true
+		delete(d.pendingReconstruct, rs.DealerIndex)
+	}
+	return nil
+}
+
+// Finished returns true if the DKG has operated the protocol correctly and has
+// all necessary information to generate the DistKeyShare() by itself. It
+// returns false otherwise.
+func (d *DistKeyGenerator) Finished() bool {
+	var ret = true
+	var nb = 0
+	d.qualIter(func(idx uint32, v *vss.Verifier) bool {
+		nb++
+		// ALL QUAL members should have their commitments by now either given or
+		// reconstructed.
+		if _, ok := d.commitments[idx]; !ok {
+			ret = false
+			return false
+		}
+		return true
+	})
+	return nb >= d.t && ret
+}
+
+// ProduceSharedSecret generates the sharedsecret relative to this receiver
+// It throws an error if something is wrong such as not enough deals received.
+// The shared secret can be computed when all deals have been sent and
+// basically consists of a
+// 1. Public point which is the sum of all aggregated individual public commits
```

Comment:
godoc: bad formatting. Either add an empty line before or indent and make it code. I think most of the time we indented it to make it look like code.